pkg/virt-launcher/virtwrap/live-migration-source.go

```diff
@@ -531,17 +536,37 @@ func (m *migrationMonitor) processInflightMigration(dom cli.VirDomain, stats *li
 		// If we were to abort the migration due to a timeout while in post copy,
 		// then it would result in that active state being lost.
 
-	case m.shouldTriggerPostCopy(elapsed):
-		logger.Info("Starting post copy mode for migration")
-		// if a migration has stalled too long, post copy will be
-		// triggered when allowPostCopy is enabled
-		err := dom.MigrateStartPostCopy(uint32(0))
-		if err != nil {
-			logger.Reason(err).Error("failed to start post migration")
-			return nil
-		}
+	case m.shouldAssistMigrationToComplete(elapsed) && !m.isPausedMigration():
 
-		m.l.updateVMIMigrationMode(v1.MigrationPostCopy)
+		if m.options.AllowPostCopy {
+			logger.Info("Starting post copy mode for migration")
+			// if a migration has stalled too long, post copy will be
+			// triggered when allowPostCopy is enabled
+			err := dom.MigrateStartPostCopy(uint32(0))
+			if err != nil {
+				logger.Reason(err).Error("failed to start post migration")
+				return nil
+			}
+
+			m.l.updateVMIMigrationMode(v1.MigrationPostCopy)
+		} else {
+
+			logger.Info("Pausing the guest to allow migration to complete")
+			// if a migration has stalled too long, the guest will be paused
+			// to complete the migration when allowPostCopy is disabled
+			err := dom.Suspend()
```

Comment:
Two questions:
1. If the migration succeeds, does it mean that the target would be "resumed" and won't remain paused, right?
2. Let's say the migration still doesn't converge (for some reason?) by the time the new `acceptableCompletionTime` is reached - what will unpause the guest?

Reply:
> Two questions:
> 
> 1. If the migration succeeds, does it mean that the target would be "resumed" and won't remain paused, right?

Yes. libvirt will reuse it on a successful migration.

> 2. Let's say the migration still doesn't converge (for some reason?) by the time the new `acceptableCompletionTime` is reached - what will unpause the guest?

`acceptableCompletionTime` doesn't play a role after it has been triggered once. Only manual cancelation can stop the process.
Either the admin or the workload owner can unpause it.

Reply:
> > 2. Let's say the migration still doesn't converge (for some reason?) by the time the new `acceptableCompletionTime` is reached - what will unpause the guest?
> 
> `acceptableCompletionTime` doesn't play a role after it has been triggered once. Only manual cancelation can stop the process. Either the admin or the workload owner can unpause it.

WDYT about resuming the guest when `acceptableCompletionTime` is reached, or IOW, when the migration is no longer active?

Reply:
It's something to think about, but surely not part of this PR.

Reply:
Looks like `ProgressTimeout` still fails the migration, should it be switched to also pause?

Reply:
If we are not making progress it means that there is no network and packets cannot empty the buffer. We should cancel instead of being indefinitely stuck.